arangod/Cluster/HeartbeatThread.cpp

```diff
@@ -1224,9 +1224,26 @@ bool HeartbeatThread::handlePlanChangeCoordinator(uint64_t currentPlanVersion) {
       return false;
     }
 
+    // get the list of databases that we know about locally
+    std::vector<TRI_voc_tick_t> localIds = databaseFeature.getDatabaseIds(false);
+    for (auto id : localIds) {
+      auto r = std::find(ids.begin(), ids.end(), id);
+
+      if (r == ids.end()) {
+        // local database not found in the plan...
+        TRI_vocbase_t* db = databaseFeature.useDatabase(id);
+        TRI_ASSERT(db);
+        std::string dbName = db ? db->name() : "n/a";
+        if (db) {
+          db->release();
```

Comment:
This is not 100% safe, as `db->release()` may not be called if the assignment to the string `dbName` before fails with an exception. Then the `db` handle will be leaked.

Reply:
are you referring to the string assignment? in any case you seem to know more here. please propose a fix.

Reply:
Assigning to an std::string may allocate memory, and this can fail with `std::bad_alloc`, always.

The fix is to make sure that we call `db->release()` if it is not a nullptr, e.g.
```
std::string dbName = "n/a";

TRI_vocbase_t* db = databaseFeature.useDatabase(id);
TRI_ASSERT(db);
if (db) {
  try {
    dbName = db->name();
  } catch (...) {
    db->release();
    throw;
  }
  db->release();
}
```
Alternatively you can use a ScopeGuard.

Reply:
sure. at which point all bets are off. but taking the suggestion